control-plane/agents/src/bin/ha/cluster/switchover.rs

```diff
@@ -257,59 +294,77 @@ impl SwitchOverEngine {
             info!("Starting worker-{} of Switchover Engine", i);
             let cloned_self = self.clone();
             let cloned_channel = recv.clone();
-            tokio::spawn(async move { cloned_self.process_request(cloned_channel, i).await });
+            tokio::spawn(async move { cloned_self.worker(cloned_channel, i).await });
         }
     }
 
     /// Switchover request to be handled synchronously in each worker task.
-    async fn process_request(
-        self,
-        recv: Arc<Mutex<UnboundedReceiver<SwitchOverRequest>>>,
-        worker_num: u8,
-    ) {
+    async fn worker(self, recv: Arc<Mutex<UnboundedReceiver<SwitchOverRequest>>>, worker_num: u8) {
         loop {
             let mut lo = recv.lock().await;
             if let Some(mut q) = lo.recv().await {
+                drop(lo);
                 info!(
-                    "switchover for vol: {} picked by worker-{}",
+                    "switchover for volume: {} picked by worker-{}",
                     q.volume_id, worker_num
                 );
-                drop(lo);
-                let retry_delay_min = std::time::Duration::from_secs(1);
-                let retry_delay_max = std::time::Duration::from_secs(384);
-                let mut retry_delay = retry_delay_min;
                 loop {
                     // TODO: handle error handling properly
                     let result = match q.stage {
                         Stage::Init => q.initialize(&self.etcd).await,
                         Stage::RepublishVolume => q.republish_volume(&self.etcd).await,
                         Stage::PublishPath => q.publish_path(&self.etcd, &self.nodes).await,
                         Stage::DeleteTarget => q.delete_target(&self.etcd).await,
+                        Stage::Errored => match q.errored_switchover(&self.etcd).await {
+                            Ok(_) => break,
+                            Err(e) => Err(e),
+                        },
                         Stage::Successful => match q.delete_switchover(&self.etcd).await {
                             Ok(_) => break,
                             Err(e) => Err(e),
                         },
-                        _ => break,
                     };
                     match result {
                         Ok(_) => {
-                            // reset retry delay back to the start.
-                            retry_delay = retry_delay_min;
+                            // reset retry count back to the start after successfully completing a
+                            // stage.
+                            q.retry_count = 0;
                         }
                         Err(_) => {
-                            retry_delay = (retry_delay * 3).min(retry_delay_max);
-                            tokio::time::sleep(retry_delay).await;
+                            q.retry_count += 1;
+                            let retry_delay = if q.retry_count < NUM_OF_FAST_REQUEUE {
```

Comment:
can we calculate the retry_delay in the requeue function itself?

Reply:
WIll do

Reply:
Incorporated